apps/internal/base/base.go

```diff
@@ -345,6 +350,24 @@ func (b Client) AcquireTokenSilent(ctx context.Context, silent AcquireTokenSilen
 	if silent.Claims == "" {
 		ar, err = AuthResultFromStorage(storageTokenResponse)
 		if err == nil {
+			if shouldRefresh(storageTokenResponse.AccessToken.RefreshOn.T) {
+				if tr, er := b.Token.Credential(ctx, authParams, silent.Credential); er == nil {
+					return b.AuthResultFromToken(ctx, authParams, tr)
+				} else if callErr, ok := er.(*errors.CallErr); ok {
+					// Check if the error is of type CallErr and matches the relevant status codes
+					switch callErr.Resp.StatusCode {
+					case http.StatusRequestTimeout, // 408
+						http.StatusTooManyRequests,     // 429
+						http.StatusInternalServerError, // 500
+						http.StatusBadGateway,          // 502
+						http.StatusServiceUnavailable,  // 503
+						http.StatusGatewayTimeout:      // 504
+					default:
+						// Handle non-retryable errors
+						return AuthResult{}, er
```

Comment:
Why not return the valid token from the cache?

Reply:
@bgavrilMS can you please explain the reasoning for this behavior? What's the benefit of MSAL returning an error to the application when it has a valid token it could return instead?